lib/MySQL_Monitor.cpp

```diff
@@ -3447,58 +3471,97 @@ VALGRIND_ENABLE_ERROR_REPORTING;
+	}
 
-			try {
-				current_discovered_port_int = stoi(s[3]);
-			} catch (...) {
-				proxy_error(
-					"Unable to parse port value coming from '%s' during topology discovery ('%s':%s). Terminating discovery early.\n",
-					originating_server_hostname.c_str(), current_discovered_hostname.c_str(), current_discovered_port_string.c_str()
-				);
-				return;
-			}
+	vector<tuple<string, uint16_t, uint32_t, int64_t, int32_t>> new_servers;
 
-			if (find(saved_hostnames.begin(), saved_hostnames.end(), current_discovered_hostname) == saved_hostnames.end()) {
-				tuple<string, int, int> new_server(current_discovered_hostname, current_discovered_port_int, reader_hostgroup);
-				new_servers.push_back(new_server);
-				saved_hostnames.push_back(current_discovered_hostname);
-			}
+	// Loop through discovered servers and process the ones we haven't saved yet
+	for (MYSQL_ROW row : discovered_servers) {
+		if ( !row ) {
+			proxy_warning("Received empty RDS topology record from %s.\n", originating_server_hostname.c_str());
+			continue;
+		}
+		int current_discovered_read_only = 1;
+VALGRIND_DISABLE_ERROR_REPORTING;
+		if (row[0]) {
+			if (!strcmp(row[0], "0") || !strcasecmp(row[0], "OFF"))
+			current_discovered_read_only = 0;
+		}
+VALGRIND_ENABLE_ERROR_REPORTING;
+		string current_discovered_hostname = row[2];
+		string current_discovered_port_string = row[3];
+		uint16_t current_discovered_port;
+		try {
+			current_discovered_port = (uint16_t)stoi(current_discovered_port_string);
+		} catch (...) {
+			proxy_error(
+				"Unable to parse port value coming from '%s' during topology discovery ('%s':%s). Terminating discovery early.\n",
+				originating_server_hostname.c_str(), current_discovered_hostname.c_str(), current_discovered_port_string.c_str()
+			);
+			return;
+		}
+		string current_discovered_role, current_discovered_status, current_discovered_version;
+		if (rds_topology_check_type == AWS_RDS_BLUE_GREEN_DEPLOYMENT_STATE_CHECK && num_fields >= 7) {
+			current_discovered_role = row[4];
+			current_discovered_status = row[5];
+			current_discovered_version = row[6];
+		}
+		if (!current_discovered_version.empty() && !is_aws_rds_topology_version_supported(current_discovered_version)) {
+			proxy_warning("Discovered topology version (%s) is not compatible with supported version (%s)\n",
+				current_discovered_version.c_str(), SUPPORTED_AWS_RDS_TOPOLOGY_VERSION);
+			return;
+		}
+
+		int64_t current_determined_weight = (int64_t)(-1L); // TODO: Add logic for selecting a different weight based on discovered role and status
+		int32_t use_ssl = 0;
+		if (mmsd->use_ssl) {
+			use_ssl = 1;
 		}
 
-		// Add the new servers if any
-		if (!new_servers.empty()) {
-			MyHGM->add_discovered_servers_to_mysql_servers_and_replication_hostgroups(new_servers);
+		tuple<string, uint16_t, uint32_t, int64_t, int32_t> discovered_server(current_discovered_hostname, current_discovered_port, reader_hostgroup, current_determined_weight, use_ssl);
```

Comment:
This is a general stylistic comment. We try to stick to a `120` char limit, it's not an strict rule that excludes debug messages, but if you could re-format some of these very long lines, that would be great.